### ISSUE TITLE
Type name references get converted to copies of type definition in generated schema

### ISSUE THREAD

**Issue by tsutsu** [actgardner/gogen-avro#69]
For example, if I have one field in my schema that looks like this:

```json
{"name": "Hash", "type": {"type": "fixed", "size": 32, "name": "sha256"}}
```

...and then another field, later on in the schema, that looks like this:

```json
{"name": "AnotherHash", "type": "com.example.sha256"}
```

...then the schema string-literal which `gogen-avro` emits as the body of the `Schema()` method for this record type, should look pretty much the same. The first occurrence of the `sha256` type should be a fully-expanded definition, and then all later occurrences should be (fully-qualified) references to the type.

Instead, though, in the generated code of `gogen-avro`, the expanded definitional form of the `sha256` type appears every time.

This isn't allowed, per the Avro spec, and causes the official Apache `avro-tools` to barf:

```
Exception in thread "main" org.apache.avro.SchemaParseException: Can't redefine: com.example.sha256
	at org.apache.avro.Schema$Names.put(Schema.java:1128)
	at org.apache.avro.Schema$Names.add(Schema.java:1123)
	at org.apache.avro.Schema.parse(Schema.java:1317)
	at org.apache.avro.Schema.parse(Schema.java:1269)
	at org.apache.avro.Schema.parse(Schema.java:1269)
	at org.apache.avro.Schema$Parser.parse(Schema.java:1032)
	at org.apache.avro.Schema$Parser.parse(Schema.java:1020)
	at org.apache.avro.Schema.parse(Schema.java:1081)
	at org.apache.avro.file.DataFileStream.initialize(DataFileStream.java:124)
	at org.apache.avro.file.DataFileStream.<init>(DataFileStream.java:84)
	at org.apache.avro.tool.CatTool.nextInput(CatTool.java:147)
	at org.apache.avro.tool.CatTool.run(CatTool.java:88)
	at org.apache.avro.tool.Main.run(Main.java:87)
	at org.apache.avro.tool.Main.main(Main.java:76)
```

**Comment by actgardner** [comment]
Thanks for filing this issue! I'll take a look, we definitely want to make sure the OCF ouptut files are compliant with the Avro spec.

**Comment by actgardner** [comment]
I've confirmed that a bug existed in the 5.0 release line and master where `fixed` and `enum` type definitions were repeated for every field. This issue should be fixed on master and in the 5.2.3 release on gopkg.in, and it doesn't affect 4.0 or earlier.

I'm going to close this issue as fixed, please feel free to open a new issue if the problem is still happening or you run into anything else!
